test/Npgsql.Tests/CopyTests.cs

```diff
@@ -696,6 +696,40 @@ public void ImportExceptionDoesNotCommit()
             }
         }
 
+        [Test, IssueLink("https://github.com/npgsql/npgsql/issues/2347")]
+        public void Write_ColumnOutOfBounds_ThrowsInvalidOperationException()
+        {
+            using (var conn = OpenConnection())
```

Comment:
Use a new C# 8 feature - [using declarations](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#using-declarations):

```suggestion
            using var conn = OpenConnection();
```

Reply:
Not sure I follow you. Do you want two new exceptions inside `ThrowHelper`? Line `117` seems a different problem to me and in my view it deserves a different exception, or? This example you gave suits well for it, but not so much for the case this PR is for:  where people call ` writer.Write` more times than allowed, thus getting an index out of bound exception.

Reply:
I deleted my previous comment because I get I rushed to a conclusion that was wrong. But:

The problem is that at the time of an `index out of bounds` exception, `_column` and `NumColumns` have the same value. So, using them with the values they have, doesn't seem to be a very meaningful error message. For example, if I start the process with 2 columns and try to add a third, I'll get this message:

`The binary import operation was started with 2 column(s), but 2 value(s) were provided`. 

We could +1 on `_column` before passing it to `ThrowHelper`, so the message makes more sense.

Reply:
Yep.